rclcpp/include/rclcpp/create_timer.hpp

```diff
@@ -102,6 +102,26 @@ create_wall_timer(
     throw std::invalid_argument{"input node_timers cannot be null"};
   }
 
+  if (period < std::chrono::duration<DurationRepT, DurationT>::zero()) {
+    throw std::invalid_argument{"timer period cannot be negative"};
+  }
+
+  // If period is greater than nanoseconds::max(), the duration_cast to nanoseconds will overflow
+  // a signed integer, which is undefined behavior. Checking whether any std::chrono::duration is
+  // greater than nanoseconds::max() is a difficult general problem. This is a more conservative
+  // version of Howard Hinnant's (the <chrono> guy>) response here:
+  // https://stackoverflow.com/a/44637334/2089061
+  // However, this doesn't solve the issue for all possible duration types of period.
+  // Follow-up issue: https://github.com/ros2/rclcpp/issues/1177
+  constexpr auto ns_max_as_double =
+    std::chrono::duration_cast<
+    std::chrono::duration<double, std::chrono::nanoseconds::period>
+    >(std::chrono::nanoseconds::max() - std::chrono::duration<DurationRepT, DurationT>(1));
```

Comment:
@brawner why substracting an `std::chrono::duration<DurationRepT, DurationT>` unit?

Reply:
The DurationRepT of nanoseconds is a 64-bit signed int, however the precision of a 64-bit double is such that the duration cast to double might actually cause period to round down to be less than nanoseconds::max(). Essentially, 

```
duration_cast<duration<double, nanoseconds::period>>(nanoseconds::max() + 1us) == duration_cast<duration<double, nanoseconds::period>>(nanoseconds::max())
```

But, this would overflow.
```
duration_cast<nanoseconds>(1us + nanoseconds::max())
```

In the stackover flow response I posted, they would cast period to a double check if it would overflow, then cast the double to nanoseconds, which also works. I just didn't want to change the behavior for the by far most common case of period << nanoseconds::max().

Reply:
Mind to update the comment to reflect that rationale?

Reply:
Added a comment and split it out into a separate variable for clarity.